server/src/main/kotlin/org/javacs/kt/externalsources/KlsURI.kt

```diff
@@ -37,16 +37,28 @@ data class KlsURI(val uri: URI) {
             .split(".")
             .takeIf { it.size > 1 }
             ?.lastOrNull()
+    val source: Boolean get() = uri.schemeSpecificPart.split("?").getOrNull(1)?.split("&")?.find {
+        it.matches("source=(true|false)".toRegex())
+    }?.split("=")?.get(1)?.toBoolean() ?: false
     val isCompiled: Boolean
         get() = fileExtension == "class"
 
     fun withFileExtension(newExtension: String): KlsURI {
-        val (parentURI, fileName) = uri.toString().partitionAroundLast("/")
-        val newURI = "$parentURI${fileName.split(".").first()}.$newExtension"
+        val (parentURI, fileNamePlusQuery) = uri.toString().partitionAroundLast("/")
+        val (fileName, query) = if (fileNamePlusQuery.contains("?")) fileNamePlusQuery.partitionAroundLast("?") else Pair(fileNamePlusQuery, "")
+        val newURI = "$parentURI${fileName.split(".").first()}.$newExtension$query"
         return KlsURI(URI(newURI))
     }
 
-    private fun toFileURI(): URI = URI(uri.schemeSpecificPart)
+    fun withSource(source: Boolean): KlsURI {
+        return KlsURI(URI("${uri.toString()}?source=$source"))
+    }
+
+    fun withoutQuery(): KlsURI {
+        return KlsURI(URI(uri.toString().split("?")[0]))
+    }
+
+    fun toFileURI(): URI = URI(uri.schemeSpecificPart)
```

Comment:
Perhaps we should use a more deconstructed representation of KLS-URIs internally, so we don't have to explicitly parse them every time. Something along the lines of

```kotlin
data class KlsURI(
    val fileURI: URI,
    val isSource: Boolean
)
```

or

```kotlin
data class KlsURI(
    val fileURI: URI,
    val query: Map<String, String>
)
```

In the latter case we might want to add constants for the known query keys.

Reply:
I think we can do that, but we'll still need to parse the URIs to fetch the source (just in different places I would say). When a request comes in to fetch the JAR contents, we would need to either fetch the source in the `URIContentProvider` and pass it to the `KlsURI` or parse the URI directly in the KlsURI constructor. And we would still need to append the source on `toString`.

Ideally, `URI` should be able to hold all this logic (and it does for normal URIs), but with KLS URIs, `URI` doesn't seem to be able to parse the query properly (I'm guessing it's because of the KLS and file schemes in the beginning).

I'm not actually sure what the best approach would be here to be honest

Reply:
I think I get your idea now. I'll see what I can do

Reply:
I updated the PR with an approach that is more in line with your idea.